core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.seqno;
+
+import com.carrotsearch.hppc.ObjectLongHashMap;
+import com.carrotsearch.hppc.ObjectLongMap;
+import com.carrotsearch.hppc.cursors.ObjectLongCursor;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.AbstractIndexShardComponent;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.Set;
+
+/**
+ * A shard component that is responsible of tracking the global checkpoint. The global checkpoint
+ * is the highest seq_no for which all lower (or equal) seq no have been process on all shards that
+ * are currently active. Since shards count as "active" when the master starts them, and before this primary shard
+ * has been notified of this fact, we also include shards in that are in the
+ * {@link org.elasticsearch.index.shard.IndexShardState#POST_RECOVERY} state when checking for global checkpoint advancement.
+ * We call these shards "in sync" with all operations on the primary (see {@link #inSyncLocalCheckpoints}.
+ *
+ * <p>
+ * The global checkpoint is maintained by the primary shard and is replicated to all the replicas
+ * (via {@link GlobalCheckpointSyncAction}).
+ */
+public class GlobalCheckpointService extends AbstractIndexShardComponent {
+
+    /**
+     * This map holds the last known local checkpoint for every shard copy that's active.
+     * All shard copies in this map participate in determining the global checkpoint
+     * keyed by allocation ids
+     */
+    final private ObjectLongMap<String> activeLocalCheckpoints;
+
+    /**
+     * This map holds the last known local checkpoint for every initializing shard copy that's has been brought up
+     * to speed through recovery. These shards are treated as valid copies and participate in determining the global
+     * checkpoint.
+     * <p>
+     * Keyed by allocation ids.
+     */
+    final private ObjectLongMap<String> inSyncLocalCheckpoints; // keyed by allocation ids
+
+
+    /**
+     * This map holds the last known local checkpoint for every initializing shard copy that is still undergoing recovery.
+     * These shards <strong>do not</strong> participate in determining the global checkpoint. This map is needed to make sure that when
+     * shards are promoted to {@link #inSyncLocalCheckpoints} we use the highest known checkpoint, even if we index concurrently
+     * while recovering the shard.
+     * Keyed by allocation ids
+     */
+    final private ObjectLongMap<String> trackingLocalCheckpoint;  // keyed by allocation ids
```

Comment:
Duplicate documentation of being keyed by allocation ids, here and in the above Javadoc.